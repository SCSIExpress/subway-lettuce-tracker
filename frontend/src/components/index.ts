// Component exports will be added as components are created
// This file serves as a central export point for all components

export { ManualLocationEntry } from './ManualLocationEntry';
export { LocationPermissionHandler } from './LocationPermissionHandler';
export { default as MapView } from './MapView';
export { default as LocationPanel } from './LocationPanel';
export { default as LocationCard } from './LocationCard';
export { default as RatingModal } from './RatingModal';
export { default as ErrorBoundary, QueryErrorBoundary, useErrorHandler } from './ErrorBoundary';
export { default as DataSyncIndicator, DataFreshnessIndicator } from './DataSyncIndicator';
export { default as OfflineIndicator, OfflineMessage, useOfflineErrorMessage } from './OfflineIndicator';