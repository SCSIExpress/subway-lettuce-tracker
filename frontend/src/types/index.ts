// Core coordinate interface
export interface Coordinates {
  lat: number;
  lng: number;
}

// Store hours interface
export interface StoreHours {
  monday: { open: string; close: string; closed?: boolean };
  tuesday: { open: string; close: string; closed?: boolean };
  wednesday: { open: string; close: string; closed?: boolean };
  thursday: { open: string; close: string; closed?: boolean };
  friday: { open: string; close: string; closed?: boolean };
  saturday: { open: string; close: string; closed?: boolean };
  sunday: { open: string; close: string; closed?: boolean };
  timezone: string;
}

// API response interfaces
export interface SubwayLocation {
  id: string;
  name: string;
  address: string;
  coordinates: Coordinates;
  hours: StoreHours;
  lettuceScore: number; // 1-5, weighted average of last 10 ratings
  lastRated?: Date;
  recentlyRated: boolean; // rated within last 2 hours
  distanceFromUser?: number; // in meters
  isOpen?: boolean; // calculated based on current time and hours
}

export interface Rating {
  id: string;
  locationId: string;
  score: number; // 1-5
  timestamp: Date;
  userId?: string; // for future user tracking
}

export interface TimeRecommendation {
  period: 'morning' | 'lunch' | 'afternoon' | 'evening';
  averageScore: number;
  confidence: 'low' | 'medium' | 'high';
  sampleSize: number;
  timeRange: string; // e.g., "6:00 AM - 11:00 AM"
}

export interface SubwayLocationDetail extends SubwayLocation {
  ratings: Rating[];
  timeRecommendations: TimeRecommendation[];
  totalRatings: number;
  averageScore: number;
}

// API request/response types
export interface NearbyLocationsRequest {
  lat: number;
  lng: number;
  radius?: number; // meters, default 5000
  limit?: number; // max results, default 20
}

export interface NearbyLocationsResponse {
  locations: SubwayLocation[];
  userLocation: Coordinates;
  searchRadius: number;
  totalFound: number;
}

export interface SubmitRatingRequest {
  locationId: string;
  score: number; // 1-5
  userId?: string; // optional for future use
}

export interface SubmitRatingResponse {
  rating: Rating;
  newLocationScore: number;
  message: string;
}

// Error response interface
export interface ApiError {
  error: string;
  message: string;
  code?: string;
  details?: any;
}

// Utility types
export type TimePeriod = 'morning' | 'lunch' | 'afternoon' | 'evening';
export type ConfidenceLevel = 'low' | 'medium' | 'high';
export type DayOfWeek = 'monday' | 'tuesday' | 'wednesday' | 'thursday' | 'friday' | 'saturday' | 'sunday';

// Frontend-specific types
export interface UserLocation {
  coordinates: Coordinates;
  accuracy?: number;
  timestamp: Date;
}

export interface MapState {
  center: Coordinates;
  zoom: number;
  selectedLocationId?: string;
}

export interface LocationPanelState {
  isOpen: boolean;
  selectedLocation?: SubwayLocation;
  locations: SubwayLocation[];
}

export interface RatingModalState {
  isOpen: boolean;
  location?: SubwayLocation;
}