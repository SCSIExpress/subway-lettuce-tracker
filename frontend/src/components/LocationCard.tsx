import React from 'react';
import { SubwayLocation } from '../types';
import { DataFreshnessIndicator } from './DataSyncIndicator';

interface LocationCardProps {
  location: SubwayLocation;
  isSelected?: boolean;
  onSelect?: (location: SubwayLocation) => void;
  onRate?: (locationId: string) => void;
  onDirections?: (location: SubwayLocation) => void;
  showActions?: boolean;
  className?: string;
  dataUpdatedAt?: Date;
  isStale?: boolean;
}

const LocationCard: React.FC<LocationCardProps> = ({
  location,
  isSelected = false,
  onSelect,
  onRate,
  onDirections,
  showActions = true,
  className = '',
  dataUpdatedAt,
  isStale = false,
}) => {
  const formatDistance = (distance?: number): string => {
    if (distance === undefined || distance === null) return '';
    if (distance < 1000) return `${Math.round(distance)}m`;
    return `${(distance / 1000).toFixed(1)}km`;
  };

  const getScoreColor = (score: number): string => {
    if (score >= 4.5) return 'bg-green-600';
    if (score >= 4) return 'bg-green-500';
    if (score >= 3.5) return 'bg-yellow-500';
    if (score >= 3) return 'bg-yellow-600';
    if (score >= 2.5) return 'bg-orange-500';
    if (score >= 2) return 'bg-orange-600';
    if (score >= 1) return 'bg-red-500';
    return 'bg-gray-400';
  };

  const getScoreTextColor = (score: number): string => {
    if (score >= 4.5) return 'text-green-700';
    if (score >= 4) return 'text-green-600';
    if (score >= 3.5) return 'text-yellow-600';
    if (score >= 3) return 'text-yellow-700';
    if (score >= 2.5) return 'text-orange-600';
    if (score >= 2) return 'text-orange-700';
    if (score >= 1) return 'text-red-600';
    return 'text-gray-500';
  };

  const renderStars = (score: number): JSX.Element[] => {
    const stars = [];
    const fullStars = Math.floor(score);
    const hasHalfStar = score % 1 >= 0.5;
    
    // Full stars
    for (let i = 0; i < fullStars; i++) {
      stars.push(
        <span key={`full-${i}`} className="text-yellow-400 text-lg">
          ★
        </span>
      );
    }
    
    // Half star
    if (hasHalfStar) {
      stars.push(
        <span key="half" className="text-yellow-400 text-lg relative">
          <span className="absolute inset-0">☆</span>
          <span className="absolute inset-0 overflow-hidden w-1/2">★</span>
        </span>
      );
    }
    
    // Empty stars
    const emptyStars = 5 - Math.ceil(score);
    for (let i = 0; i < emptyStars; i++) {
      stars.push(
        <span key={`empty-${i}`} className="text-gray-300 text-lg">
          ☆
        </span>
      );
    }
    
    return stars;
  };

  const getCurrentStatus = (): { isOpen: boolean; text: string; color: string } => {
    const now = new Date();
    const dayNames = ['sunday', 'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday'];
    const currentDay = dayNames[now.getDay()] as keyof typeof location.hours;
    const currentTime = now.getHours() * 100 + now.getMinutes();
    
    const todayHours = location.hours[currentDay];
    
    // Handle case where hours might be undefined or in wrong format
    if (typeof todayHours === 'string' || !todayHours) {
      return { isOpen: false, text: 'Hours unavailable', color: 'text-gray-500' };
    }
    
    if (todayHours.closed) {
      return { isOpen: false, text: 'Closed today', color: 'text-red-600' };
    }

    // Parse time strings (assuming format like "09:00" or "9:00")
    const parseTime = (timeStr: string): number => {
      const [hours, minutes] = timeStr.split(':').map(Number);
      return hours * 100 + (minutes || 0);
    };

    const openTime = parseTime(todayHours.open);
    const closeTime = parseTime(todayHours.close);
    
    const isOpen = currentTime >= openTime && currentTime <= closeTime;
    
    if (isOpen) {
      return { 
        isOpen: true, 
        text: `Open until ${todayHours.close}`, 
        color: 'text-green-600' 
      };
    } else if (currentTime < openTime) {
      return { 
        isOpen: false, 
        text: `Opens at ${todayHours.open}`, 
        color: 'text-orange-600' 
      };
    } else {
      return { 
        isOpen: false, 
        text: 'Closed', 
        color: 'text-red-600' 
      };
    }
  };

  const formatLastRated = (lastRated?: Date): string => {
    if (!lastRated) return '';
    
    const now = new Date();
    const diffMs = now.getTime() - new Date(lastRated).getTime();
    const diffHours = Math.floor(diffMs / (1000 * 60 * 60));
    const diffDays = Math.floor(diffHours / 24);
    
    if (diffHours < 1) return 'Just rated';
    if (diffHours < 24) return `${diffHours}h ago`;
    if (diffDays < 7) return `${diffDays}d ago`;
    return new Date(lastRated).toLocaleDateString();
  };

  const status = getCurrentStatus();
  const hasValidScore = location.lettuceScore > 0;

  const handleCardClick = () => {
    if (onSelect) {
      onSelect(location);
    }
  };

  const handleRateClick = (e: React.MouseEvent) => {
    e.stopPropagation();
    if (onRate) {
      onRate(location.id);
    }
  };

  const handleDirectionsClick = (e: React.MouseEvent) => {
    e.stopPropagation();
    if (onDirections) {
      onDirections(location);
    }
  };

  return (
    <div
      className={`bg-white rounded-xl sm:rounded-2xl shadow-md hover:shadow-lg transition-all duration-200 tap-target ${
        isSelected ? 'ring-2 ring-green-500 shadow-lg' : ''
      } ${onSelect ? 'cursor-pointer' : ''} ${className}`}
      onClick={handleCardClick}
      role={onSelect ? 'button' : 'article'}
      tabIndex={onSelect ? 0 : undefined}
      onKeyDown={onSelect ? (e) => {
        if (e.key === 'Enter' || e.key === ' ') {
          e.preventDefault();
          handleCardClick();
        }
      } : undefined}
      aria-label={onSelect ? `Select ${location.name}` : undefined}
    >
      <div className="spacing-mobile">
        {/* Header with name, distance, and recently rated indicator */}
        <div className="flex justify-between items-start mb-3 sm:mb-4">
          <div className="flex-1 min-w-0">
            <div className="flex items-center gap-2 mb-1 flex-wrap">
              <h3 className="font-semibold text-gray-900 text-responsive-base leading-tight truncate">
                {location.name}
              </h3>
              {location.recentlyRated && (
                <div 
                  className="flex items-center gap-1 bg-green-100 px-2 py-1 rounded-full flex-shrink-0"
                  title="Recently rated within the last 2 hours"
                >
                  <div className="w-2 h-2 bg-green-500 rounded-full animate-pulse"></div>
                  <span className="text-xs text-green-700 font-medium whitespace-nowrap">
                    Fresh
                  </span>
                </div>
              )}
            </div>
            <p className="text-gray-600 text-responsive-xs leading-relaxed">
              {location.address}
            </p>
          </div>
          
          {location.distanceFromUser !== undefined && location.distanceFromUser !== null && (
            <div className="text-right ml-2 sm:ml-3 flex-shrink-0">
              <span className="text-xs sm:text-sm font-medium text-gray-700 bg-gray-100 px-2 py-1 rounded-lg">
                {formatDistance(location.distanceFromUser)}
              </span>
            </div>
          )}
        </div>

        {/* Score display with stars and colors */}
        <div className="flex items-center justify-between mb-3 sm:mb-4">
          <div className="flex items-center gap-2 sm:gap-3 min-w-0 flex-1">
            {/* Numerical score badge */}
            <div className="flex items-center gap-2 min-w-0">
              <div 
                className={`w-9 h-9 sm:w-10 sm:h-10 rounded-full flex items-center justify-center text-white text-xs sm:text-sm font-bold shadow-sm flex-shrink-0 ${
                  hasValidScore ? getScoreColor(location.lettuceScore) : 'bg-gray-400'
                }`}
                title={hasValidScore ? `Lettuce freshness score: ${location.lettuceScore.toFixed(1)}/5` : 'No ratings yet'}
              >
                {hasValidScore ? location.lettuceScore.toFixed(1) : 'N/A'}
              </div>
              
              {/* Star rating */}
              <div className="flex items-center gap-0.5 min-w-0">
                {hasValidScore ? (
                  <div className="flex items-center gap-0.5">
                    {renderStars(location.lettuceScore).map((star, index) => (
                      <span key={index} className="text-sm sm:text-lg">
                        {star}
                      </span>
                    ))}
                  </div>
                ) : (
                  <span className="text-gray-400 text-responsive-xs">No ratings</span>
                )}
              </div>
            </div>

            {/* Score text description */}
            {hasValidScore && (
              <div className="hidden md:block flex-shrink-0">
                <span className={`text-responsive-xs font-medium ${getScoreTextColor(location.lettuceScore)}`}>
                  {location.lettuceScore >= 4.5 ? 'Excellent' :
                   location.lettuceScore >= 4 ? 'Very Good' :
                   location.lettuceScore >= 3.5 ? 'Good' :
                   location.lettuceScore >= 3 ? 'Fair' :
                   location.lettuceScore >= 2 ? 'Poor' : 'Very Poor'}
                </span>
              </div>
            )}
          </div>

          {/* Open/closed status */}
          <div className="text-right flex-shrink-0 ml-2">
            <div className="flex items-center gap-1">
              <div className={`w-2 h-2 rounded-full ${status.isOpen ? 'bg-green-500' : 'bg-red-500'}`}></div>
              <span className={`text-xs font-medium ${status.color} whitespace-nowrap`}>
                {status.text}
              </span>
            </div>
          </div>
        </div>

        {/* Last rated information and data freshness */}
        <div className="mb-3 sm:mb-4 flex items-center justify-between flex-wrap gap-2">
          {location.lastRated && (
            <span className="text-responsive-xs text-gray-500">
              Last rated: {formatLastRated(location.lastRated)}
            </span>
          )}
          <DataFreshnessIndicator 
            lastUpdated={dataUpdatedAt}
            isStale={isStale}
            className="ml-auto flex-shrink-0"
          />
        </div>

        {/* Action buttons - show when selected or always if showActions is true */}
        {showActions && (isSelected || !onSelect) && (
          <div className="flex flex-col sm:flex-row gap-2 sm:gap-3 pt-3 sm:pt-4 border-t border-gray-100">
            {onRate && (
              <button
                onClick={handleRateClick}
                className="flex-1 bg-green-600 hover:bg-green-700 active:bg-green-800 text-white py-3 sm:py-2.5 px-4 rounded-xl text-responsive-xs font-medium transition-colors touch-target tap-target focus:outline-none focus:ring-2 focus:ring-green-500 focus:ring-offset-2"
                aria-label={`Rate lettuce freshness at ${location.name}`}
              >
                <span className="flex items-center justify-center gap-1.5 sm:gap-1">
                  <span>Rate Lettuce</span>
                  <span className="text-base sm:text-lg">🥬</span>
                </span>
              </button>
            )}
            {onDirections && (
              <button
                onClick={handleDirectionsClick}
                className="flex-1 bg-blue-600 hover:bg-blue-700 active:bg-blue-800 text-white py-3 sm:py-2.5 px-4 rounded-xl text-responsive-xs font-medium transition-colors touch-target tap-target focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2"
                aria-label={`Get directions to ${location.name}`}
              >
                <span className="flex items-center justify-center gap-1.5 sm:gap-1">
                  <span>Directions</span>
                  <span className="text-base sm:text-lg">🧭</span>
                </span>
              </button>
            )}
          </div>
        )}
      </div>
    </div>
  );
};

export default LocationCard;