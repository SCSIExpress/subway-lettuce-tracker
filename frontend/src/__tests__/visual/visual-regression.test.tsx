import { render } from '@testing-library/react';
import { vi } from 'vitest';
import { LocationCard } from '../../components/LocationCard';
import { LocationPanel } from '../../components/LocationPanel';
import { RatingModal } from '../../components/RatingModal';
import { MapView } from '../../components/MapView';
import { LocationPermissionHandler } from '../../components/LocationPermissionHandler';
import { ManualLocationEntry } from '../../components/ManualLocationEntry';
import { SubwayLocation } from '../../types';

// Mock Google Maps
const mockMap = {
  setCenter: vi.fn(),
  setZoom: vi.fn(),
  addListener: vi.fn(),
};

const mockMarker = {
  setPosition: vi.fn(),
  setMap: vi.fn(),
  addListener: vi.fn(),
};

Object.defineProperty(window, 'google', {
  value: {
    maps: {
      Map: vi.fn(() => mockMap),
      Marker: vi.fn(() => mockMarker),
      InfoWindow: vi.fn(() => ({
        setContent: vi.fn(),
        open: vi.fn(),
        close: vi.fn(),
      })),
      LatLng: vi.fn((lat, lng) => ({ lat: () => lat, lng: () => lng })),
      MarkerClusterer: vi.fn(),
    },
  },
  writable: true,
});

// Mock location context
vi.mock('../../contexts/LocationContext', () => ({
  useLocation: () => ({
    state: {
      userLocation: { lat: 40.7128, lng: -74.0060 },
      permissionState: 'granted',
      isLoading: false,
      error: null,
    },
    actions: {
      requestLocation: vi.fn(),
      requestPermission: vi.fn(),
      setManualLocation: vi.fn(),
      getCurrentLocation: () => ({ lat: 40.7128, lng: -74.0060 }),
    },
  }),
}));

const mockLocation: SubwayLocation = {
  id: '1',
  name: 'Subway - Times Square',
  address: '1560 Broadway, New York, NY 10036',
  coordinates: { lat: 40.7589, lng: -73.9851 },
  hours: {
    monday: { open: '06:00', close: '23:00' },
    tuesday: { open: '06:00', close: '23:00' },
    wednesday: { open: '06:00', close: '23:00' },
    thursday: { open: '06:00', close: '23:00' },
    friday: { open: '06:00', close: '23:00' },
    saturday: { open: '07:00', close: '23:00' },
    sunday: { open: '08:00', close: '22:00' },
    timezone: 'America/New_York',
  },
  lettuceScore: 4.2,
  recentlyRated: true,
  distanceFromUser: 250,
  isOpen: true,
};

const mockLocations: SubwayLocation[] = [
  mockLocation,
  {
    ...mockLocation,
    id: '2',
    name: 'Subway - Penn Station',
    address: '2 Penn Plaza, New York, NY 10121',
    lettuceScore: 3.8,
    recentlyRated: false,
    distanceFromUser: 180,
  },
  {
    ...mockLocation,
    id: '3',
    name: 'Subway - Union Square',
    address: '4 Union Square S, New York, NY 10003',
    lettuceScore: 2.9,
    recentlyRated: false,
    distanceFromUser: 320,
  },
];

describe('Visual Regression Tests', () => {
  describe('LocationCard Component', () => {
    it('should render location card with high score', () => {
      const { container } = render(
        <LocationCard
          location={mockLocation}
          isSelected={false}
          onSelect={vi.fn()}
          onRate={vi.fn()}
          onDirections={vi.fn()}
        />
      );
      
      expect(container.firstChild).toMatchSnapshot('location-card-high-score');
    });

    it('should render location card with low score', () => {
      const lowScoreLocation = { ...mockLocation, lettuceScore: 2.1, recentlyRated: false };
      
      const { container } = render(
        <LocationCard
          location={lowScoreLocation}
          isSelected={false}
          onSelect={vi.fn()}
          onRate={vi.fn()}
          onDirections={vi.fn()}
        />
      );
      
      expect(container.firstChild).toMatchSnapshot('location-card-low-score');
    });

    it('should render selected location card', () => {
      const { container } = render(
        <LocationCard
          location={mockLocation}
          isSelected={true}
          onSelect={vi.fn()}
          onRate={vi.fn()}
          onDirections={vi.fn()}
        />
      );
      
      expect(container.firstChild).toMatchSnapshot('location-card-selected');
    });

    it('should render closed location card', () => {
      const closedLocation = { ...mockLocation, isOpen: false };
      
      const { container } = render(
        <LocationCard
          location={closedLocation}
          isSelected={false}
          onSelect={vi.fn()}
          onRate={vi.fn()}
          onDirections={vi.fn()}
        />
      );
      
      expect(container.firstChild).toMatchSnapshot('location-card-closed');
    });
  });

  describe('LocationPanel Component', () => {
    it('should render location panel in closed state', () => {
      const { container } = render(
        <LocationPanel
          locations={mockLocations}
          selectedLocation={undefined}
          onLocationSelect={vi.fn()}
          onRate={vi.fn()}
          onDirections={vi.fn()}
          isOpen={false}
          onToggle={vi.fn()}
        />
      );
      
      expect(container.firstChild).toMatchSnapshot('location-panel-closed');
    });

    it('should render location panel in open state', () => {
      const { container } = render(
        <LocationPanel
          locations={mockLocations}
          selectedLocation={mockLocation}
          onLocationSelect={vi.fn()}
          onRate={vi.fn()}
          onDirections={vi.fn()}
          isOpen={true}
          onToggle={vi.fn()}
        />
      );
      
      expect(container.firstChild).toMatchSnapshot('location-panel-open');
    });

    it('should render location panel with loading state', () => {
      const { container } = render(
        <LocationPanel
          locations={[]}
          selectedLocation={undefined}
          onLocationSelect={vi.fn()}
          onRate={vi.fn()}
          onDirections={vi.fn()}
          isOpen={true}
          onToggle={vi.fn()}
          isRefetching={true}
        />
      );
      
      expect(container.firstChild).toMatchSnapshot('location-panel-loading');
    });

    it('should render location panel with error state', () => {
      const { container } = render(
        <LocationPanel
          locations={[]}
          selectedLocation={undefined}
          onLocationSelect={vi.fn()}
          onRate={vi.fn()}
          onDirections={vi.fn()}
          isOpen={true}
          onToggle={vi.fn()}
          error={new Error('Failed to load locations')}
          onRetry={vi.fn()}
        />
      );
      
      expect(container.firstChild).toMatchSnapshot('location-panel-error');
    });
  });

  describe('RatingModal Component', () => {
    it('should render rating modal', () => {
      const { container } = render(
        <RatingModal
          location={mockLocation}
          isOpen={true}
          onClose={vi.fn()}
          onSubmit={vi.fn()}
        />
      );
      
      expect(container.firstChild).toMatchSnapshot('rating-modal-open');
    });

    it('should render rating modal with optimal times', () => {
      const locationWithTimes = {
        ...mockLocation,
        timeRecommendations: [
          {
            period: 'morning' as const,
            averageScore: 4.5,
            confidence: 'high' as const,
            sampleSize: 15,
            timeRange: '6:00 AM - 11:00 AM',
          },
          {
            period: 'lunch' as const,
            averageScore: 3.8,
            confidence: 'medium' as const,
            sampleSize: 8,
            timeRange: '11:00 AM - 3:00 PM',
          },
        ],
      };
      
      const { container } = render(
        <RatingModal
          location={locationWithTimes}
          isOpen={true}
          onClose={vi.fn()}
          onSubmit={vi.fn()}
        />
      );
      
      expect(container.firstChild).toMatchSnapshot('rating-modal-with-times');
    });

    it('should not render when closed', () => {
      const { container } = render(
        <RatingModal
          location={mockLocation}
          isOpen={false}
          onClose={vi.fn()}
          onSubmit={vi.fn()}
        />
      );
      
      expect(container.firstChild).toMatchSnapshot('rating-modal-closed');
    });
  });

  describe('LocationPermissionHandler Component', () => {
    it('should render location permission request', () => {
      const { container } = render(
        <LocationPermissionHandler>
          {(location, isLoading) => (
            location ? <div>Has location</div> : <div>No location</div>
          )}
        </LocationPermissionHandler>
      );
      
      expect(container.firstChild).toMatchSnapshot('location-permission-request');
    });

    it('should render loading state', () => {
      // Mock loading state
      vi.mocked(require('../../contexts/LocationContext').useLocation).mockReturnValue({
        state: {
          userLocation: null,
          permissionState: 'granted',
          isLoading: true,
          error: null,
        },
        actions: {
          requestLocation: vi.fn(),
          requestPermission: vi.fn(),
          setManualLocation: vi.fn(),
          getCurrentLocation: () => null,
        },
      });

      const { container } = render(
        <LocationPermissionHandler>
          {(location, isLoading) => (
            isLoading ? <div>Loading...</div> : <div>Not loading</div>
          )}
        </LocationPermissionHandler>
      );
      
      expect(container.firstChild).toMatchSnapshot('location-permission-loading');
    });

    it('should render permission denied state', () => {
      vi.mocked(require('../../contexts/LocationContext').useLocation).mockReturnValue({
        state: {
          userLocation: null,
          permissionState: 'denied',
          isLoading: false,
          error: { type: 'PERMISSION_DENIED', message: 'Permission denied' },
        },
        actions: {
          requestLocation: vi.fn(),
          requestPermission: vi.fn(),
          setManualLocation: vi.fn(),
          getCurrentLocation: () => null,
        },
      });

      const { container } = render(
        <LocationPermissionHandler>
          {(location, isLoading) => (
            location ? <div>Has location</div> : <div>No location</div>
          )}
        </LocationPermissionHandler>
      );
      
      expect(container.firstChild).toMatchSnapshot('location-permission-denied');
    });
  });

  describe('ManualLocationEntry Component', () => {
    it('should render manual location entry modal', () => {
      const { container } = render(
        <ManualLocationEntry
          isOpen={true}
          onLocationSet={vi.fn()}
          onCancel={vi.fn()}
        />
      );
      
      expect(container.firstChild).toMatchSnapshot('manual-location-entry-open');
    });

    it('should render coordinates tab', () => {
      const { container, rerender } = render(
        <ManualLocationEntry
          isOpen={true}
          onLocationSet={vi.fn()}
          onCancel={vi.fn()}
        />
      );
      
      // Click coordinates tab (simulate state change)
      rerender(
        <ManualLocationEntry
          isOpen={true}
          onLocationSet={vi.fn()}
          onCancel={vi.fn()}
        />
      );
      
      expect(container.firstChild).toMatchSnapshot('manual-location-entry-coordinates');
    });

    it('should not render when closed', () => {
      const { container } = render(
        <ManualLocationEntry
          isOpen={false}
          onLocationSet={vi.fn()}
          onCancel={vi.fn()}
        />
      );
      
      expect(container.firstChild).toMatchSnapshot('manual-location-entry-closed');
    });
  });

  describe('Responsive Design Snapshots', () => {
    beforeEach(() => {
      // Mock different viewport sizes
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 375,
      });
      Object.defineProperty(window, 'innerHeight', {
        writable: true,
        configurable: true,
        value: 667,
      });
    });

    it('should render location card on mobile', () => {
      const { container } = render(
        <div style={{ width: '375px' }}>
          <LocationCard
            location={mockLocation}
            isSelected={false}
            onSelect={vi.fn()}
            onRate={vi.fn()}
            onDirections={vi.fn()}
          />
        </div>
      );
      
      expect(container.firstChild).toMatchSnapshot('location-card-mobile');
    });

    it('should render location panel on mobile', () => {
      const { container } = render(
        <div style={{ width: '375px', height: '667px' }}>
          <LocationPanel
            locations={mockLocations}
            selectedLocation={mockLocation}
            onLocationSelect={vi.fn()}
            onRate={vi.fn()}
            onDirections={vi.fn()}
            isOpen={true}
            onToggle={vi.fn()}
          />
        </div>
      );
      
      expect(container.firstChild).toMatchSnapshot('location-panel-mobile');
    });

    it('should render rating modal on mobile', () => {
      const { container } = render(
        <div style={{ width: '375px', height: '667px' }}>
          <RatingModal
            location={mockLocation}
            isOpen={true}
            onClose={vi.fn()}
            onSubmit={vi.fn()}
          />
        </div>
      );
      
      expect(container.firstChild).toMatchSnapshot('rating-modal-mobile');
    });
  });

  describe('Dark Mode Snapshots', () => {
    beforeEach(() => {
      // Add dark mode class to document
      document.documentElement.classList.add('dark');
    });

    afterEach(() => {
      document.documentElement.classList.remove('dark');
    });

    it('should render location card in dark mode', () => {
      const { container } = render(
        <div className="dark">
          <LocationCard
            location={mockLocation}
            isSelected={false}
            onSelect={vi.fn()}
            onRate={vi.fn()}
            onDirections={vi.fn()}
          />
        </div>
      );
      
      expect(container.firstChild).toMatchSnapshot('location-card-dark-mode');
    });

    it('should render location panel in dark mode', () => {
      const { container } = render(
        <div className="dark">
          <LocationPanel
            locations={mockLocations}
            selectedLocation={mockLocation}
            onLocationSelect={vi.fn()}
            onRate={vi.fn()}
            onDirections={vi.fn()}
            isOpen={true}
            onToggle={vi.fn()}
          />
        </div>
      );
      
      expect(container.firstChild).toMatchSnapshot('location-panel-dark-mode');
    });
  });
});