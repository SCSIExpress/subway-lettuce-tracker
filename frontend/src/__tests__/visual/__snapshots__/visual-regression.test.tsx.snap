// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Visual Regression Tests > LocationPermissionHandler Component > should render location permission request > location-permission-request 1`] = `
<div>
  Has location
</div>
`;

exports[`Visual Regression Tests > ManualLocationEntry Component > should not render when closed > manual-location-entry-closed 1`] = `null`;

exports[`Visual Regression Tests > ManualLocationEntry Component > should render coordinates tab > manual-location-entry-coordinates 1`] = `
<div
  class="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center p-4 z-50"
>
  <div
    class="bg-white rounded-2xl shadow-xl max-w-md w-full max-h-[90vh] overflow-y-auto"
  >
    <div
      class="p-6 border-b border-gray-200"
    >
      <div
        class="flex items-center justify-between"
      >
        <h2
          class="text-xl font-semibold text-gray-800"
        >
          Enter Your Location
        </h2>
        <button
          aria-label="Close"
          class="text-gray-400 hover:text-gray-600 transition-colors"
        >
          <svg
            class="w-6 h-6"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 18L18 6M6 6l12 12"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
      <p
        class="text-sm text-gray-600 mt-2"
      >
        We need your location to find nearby Subway restaurants
      </p>
    </div>
    <div
      class="flex border-b border-gray-200"
    >
      <button
        class="flex-1 py-3 px-4 text-sm font-medium transition-colors text-green-600 border-b-2 border-green-600 bg-green-50"
      >
        📍 Address
      </button>
      <button
        class="flex-1 py-3 px-4 text-sm font-medium transition-colors text-gray-500 hover:text-gray-700"
      >
        🌐 Coordinates
      </button>
    </div>
    <div
      class="p-6"
    >
      <form
        class="space-y-4"
      >
        <div>
          <label
            class="block text-sm font-medium text-gray-700 mb-2"
            for="address"
          >
            Street Address or City
          </label>
          <input
            class="w-full px-3 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-green-500 focus:border-green-500 transition-colors"
            id="address"
            placeholder="e.g., 123 Main St, New York, NY"
            type="text"
            value=""
          />
        </div>
        <div
          class="flex gap-3"
        >
          <button
            class="flex-1 py-2 px-4 border border-gray-300 text-gray-700 rounded-lg hover:bg-gray-50 transition-colors"
            type="button"
          >
            Cancel
          </button>
          <button
            class="flex-1 py-2 px-4 bg-green-600 text-white rounded-lg hover:bg-green-700 disabled:bg-gray-300 disabled:cursor-not-allowed transition-colors"
            disabled=""
            type="submit"
          >
            Find Location
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`Visual Regression Tests > ManualLocationEntry Component > should render manual location entry modal > manual-location-entry-open 1`] = `
<div
  class="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center p-4 z-50"
>
  <div
    class="bg-white rounded-2xl shadow-xl max-w-md w-full max-h-[90vh] overflow-y-auto"
  >
    <div
      class="p-6 border-b border-gray-200"
    >
      <div
        class="flex items-center justify-between"
      >
        <h2
          class="text-xl font-semibold text-gray-800"
        >
          Enter Your Location
        </h2>
        <button
          aria-label="Close"
          class="text-gray-400 hover:text-gray-600 transition-colors"
        >
          <svg
            class="w-6 h-6"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 18L18 6M6 6l12 12"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
      <p
        class="text-sm text-gray-600 mt-2"
      >
        We need your location to find nearby Subway restaurants
      </p>
    </div>
    <div
      class="flex border-b border-gray-200"
    >
      <button
        class="flex-1 py-3 px-4 text-sm font-medium transition-colors text-green-600 border-b-2 border-green-600 bg-green-50"
      >
        📍 Address
      </button>
      <button
        class="flex-1 py-3 px-4 text-sm font-medium transition-colors text-gray-500 hover:text-gray-700"
      >
        🌐 Coordinates
      </button>
    </div>
    <div
      class="p-6"
    >
      <form
        class="space-y-4"
      >
        <div>
          <label
            class="block text-sm font-medium text-gray-700 mb-2"
            for="address"
          >
            Street Address or City
          </label>
          <input
            class="w-full px-3 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-green-500 focus:border-green-500 transition-colors"
            id="address"
            placeholder="e.g., 123 Main St, New York, NY"
            type="text"
            value=""
          />
        </div>
        <div
          class="flex gap-3"
        >
          <button
            class="flex-1 py-2 px-4 border border-gray-300 text-gray-700 rounded-lg hover:bg-gray-50 transition-colors"
            type="button"
          >
            Cancel
          </button>
          <button
            class="flex-1 py-2 px-4 bg-green-600 text-white rounded-lg hover:bg-green-700 disabled:bg-gray-300 disabled:cursor-not-allowed transition-colors"
            disabled=""
            type="submit"
          >
            Find Location
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
